File: src/main/java/me/miki/shindo/management/mods/impl/InternalSettingsMod.java
package me.miki.shindo.management.mods.impl;

import me.miki.shindo.management.nanovg.font.LegacyIcon;
import lombok.Getter;
import me.miki.shindo.Shindo;
import me.miki.shindo.injection.interfaces.IMixinMinecraft;
import me.miki.shindo.logger.ShindoLogger;
import me.miki.shindo.management.event.EventTarget;
import me.miki.shindo.management.event.impl.EventKey;
import me.miki.shindo.management.event.impl.EventPreRenderTick;
import me.miki.shindo.management.event.impl.EventToggleFullscreen;
import me.miki.shindo.management.language.TranslateText;
import me.miki.shindo.management.mods.Mod;
import me.miki.shindo.management.mods.ModCategory;
import me.miki.shindo.management.settings.impl.*;
import me.miki.shindo.management.settings.metadata.SettingRegistry;
import me.miki.shindo.gui.modmenu.category.impl.shared.SettingsPanel;
import me.miki.shindo.management.screenshot.ScreenshotDisplayMode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.lwjgl.LWJGLException;
import org.lwjgl.input.Keyboard;
import org.lwjgl.opengl.Display;
import org.lwjgl.opengl.DisplayMode;

import me.miki.shindo.management.settings.config.Property;
import me.miki.shindo.management.settings.config.PropertyEnum;
import me.miki.shindo.management.settings.config.PropertyType;

import java.util.Objects;

public class InternalSettingsMod extends Mod {

    private static InternalSettingsMod instance;

    @Property(type = PropertyType.COMBO, translate = TranslateText.HUD_THEME)
    private HudTheme hudTheme = HudTheme.NORMAL;

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.UI_BLUR)
    private boolean blurSetting = true;

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.MC_FONT)
    private boolean mcFontSetting = false;

    @Property(type = PropertyType.NUMBER, translate = TranslateText.VOLUME, min = 0, max = 1, current = 0.8)
    private double volumeSetting = 0.8;

    @Property(type = PropertyType.KEYBIND, translate = TranslateText.KEYBIND, keyCode = Keyboard.KEY_RSHIFT)
    private int modMenuKeybindSetting = Keyboard.KEY_RSHIFT;

    @Property(type = PropertyType.TEXT, translate = TranslateText.CUSTOM_CAPE, text = "None")
    private String capeNameSetting = "None";

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.CLICK_EFFECT)
    private boolean clickEffectsSetting = true;

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.UI_SOUNDS)
    private boolean soundsUISetting = true;

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.BORDERLESS_FULSCREEN)
    private boolean borderlessFullscreenSetting = false;

    @Property(type = PropertyType.COMBO, name = "Settings Layout")
    private SettingsLayout settingsLayout = SettingsLayout.SINGLE_COLUMN;

    @Property(type = PropertyType.COMBO, name = "Module Layout")
    private ModuleLayout moduleLayout = ModuleLayout.SINGLE_COLUMN;

    @Property(type = PropertyType.COMBO, name = "Addon Layout")
    private AddonLayout addonLayout = AddonLayout.STANDARD;

    @Property(type = PropertyType.COMBO, name = "Screenshot Display")
    private ScreenshotDisplayMode screenshotDisplayMode = ScreenshotDisplayMode.FILMSTRIP;

    @Property(type = PropertyType.COMBO, translate = TranslateText.NOTIFICATION_POSITION)
    private NotificationCorner notificationCorner = NotificationCorner.BOTTOM_RIGHT;

    private int prevX, prevY, prevWidth, prevHeight;
    private long fullscreenTime = -1;
    private boolean lastBorderlessState = false;
    private boolean borderlessInitialized = false;

    public InternalSettingsMod() {
        super(TranslateText.NONE, TranslateText.NONE, ModCategory.OTHER, LegacyIcon.MOD_INTERNAL_SETTINGS);

        instance = this;
    }

    @Override
    public void setup() {
        this.setHide(true);
        this.setToggled(true);
    }

    @EventTarget
    public void onKey(EventKey event) {
        if (event.getKeyCode() == modMenuKeybindSetting) {
            mc.displayGuiScreen(Shindo.getInstance().getShindoAPI().getModMenu());
        }

        // Uncomment to enable the ability to change the theme of the mod menu using the down arrow key
        if (event.getKeyCode() == Keyboard.KEY_DOWN) {
            ComboSetting combo = getModThemeSetting();
            if (combo != null) {
                int max = combo.getOptions().size();
                int modeIndex = combo.getOptions().indexOf(combo.getOption());
                modeIndex = modeIndex > 0 ? modeIndex - 1 : max - 1;
                combo.setOption(combo.getOptions().get(modeIndex));
            }
        }
    }

    @EventTarget
    public void onRenderTick(EventPreRenderTick event) {
        if (fullscreenTime != -1 && System.currentTimeMillis() - fullscreenTime >= 100) {
            fullscreenTime = -1;

            if (mc.inGameHasFocus) {
                mc.mouseHelper.grabMouseCursor();
            }
        }

        if (borderlessInitialized && borderlessFullscreenSetting != lastBorderlessState) {
            applyBorderlessSetting(borderlessFullscreenSetting, false);
        }
    }

    @EventTarget
    public void onFullscreenToggle(EventToggleFullscreen event) {
        if (!borderlessFullscreenSetting) {
            return;
        }
        event.setApplyState(false);
        setBorderlessFullscreen(event.getState());
    }

    public static InternalSettingsMod getInstance() {
        return instance;
    }

    public HudTheme getHudTheme() {
        return hudTheme;
    }

    public ScreenshotDisplayMode getScreenshotDisplayMode() {
        return screenshotDisplayMode;
    }

    public NotificationCorner getNotificationCorner() {
        return notificationCorner;
    }

    public ModuleLayout getModuleLayout() {
        return moduleLayout;
    }

    public AddonLayout getAddonLayout() {
        return addonLayout;
    }

    public SettingsLayout getSettingsLayout() {
        return settingsLayout;
    }

    public BooleanSetting getClickEffectsSetting() {
        return SettingRegistry.getBooleanSetting(this, "clickEffectsSetting");
    }

    public BooleanSetting getSoundsUISetting() {
        return SettingRegistry.getBooleanSetting(this, "soundsUISetting");
    }

    public ComboSetting getSettingsLayoutSetting() {
        return SettingRegistry.getComboSetting(this, "settingsLayout");
    }

    public ComboSetting getModuleLayoutSetting() {
        return SettingRegistry.getComboSetting(this, "moduleLayout");
    }

    public ComboSetting getAddonLayoutSetting() {
        return SettingRegistry.getComboSetting(this, "addonLayout");
    }

    public ComboSetting getScreenshotDisplaySetting() {
        return SettingRegistry.getComboSetting(this, "screenshotDisplayMode");
    }

    public ComboSetting getNotificationCornerSetting() {
        return SettingRegistry.getComboSetting(this, "notificationCorner");
    }

    public SettingsPanel.LayoutMode getSettingsLayoutMode() {
        return settingsLayout == SettingsLayout.COMPACT_GRID
                ? SettingsPanel.LayoutMode.DOUBLE_COLUMN
                : SettingsPanel.LayoutMode.SINGLE_COLUMN;
    }

    public void setSettingsLayoutMode(SettingsPanel.LayoutMode mode) {
        SettingsLayout target = mode == SettingsPanel.LayoutMode.DOUBLE_COLUMN
                ? SettingsLayout.COMPACT_GRID
                : SettingsLayout.SINGLE_COLUMN;
        ComboSetting combo = getSettingsLayoutSetting();
        if (combo != null && target.ordinal() < combo.getOptions().size()) {
            combo.setOption(combo.getOptions().get(target.ordinal()));
        }
    }

    public int getModuleGridColumns() {
        if (Objects.requireNonNull(moduleLayout) == ModuleLayout.TWO_COLUMNS) {
            return 2;
        }
        if (moduleLayout == ModuleLayout.ICON_CARDS) {
            return 3;
        }
        return 1;
    }

    public void setModuleGridColumns(int columns) {
        int normalized = Math.max(1, Math.min(columns, 2));
        ModuleLayout target = normalized == 2 ? ModuleLayout.TWO_COLUMNS : ModuleLayout.SINGLE_COLUMN;
        ComboSetting combo = getModuleLayoutSetting();
        if (combo != null && target.ordinal() < combo.getOptions().size()) {
            combo.setOption(combo.getOptions().get(target.ordinal()));
        }
    }

    public void setModuleLayout(ModuleLayout layout) {
        ModuleLayout target = layout == null ? ModuleLayout.SINGLE_COLUMN : layout;
        ComboSetting combo = getModuleLayoutSetting();
        if (combo != null && target.ordinal() < combo.getOptions().size()) {
            combo.setOption(combo.getOptions().get(target.ordinal()));
        }
    }

    public void setAddonLayout(AddonLayout layout) {
        AddonLayout target = layout == null ? AddonLayout.STANDARD : layout;
        ComboSetting combo = getAddonLayoutSetting();
        if (combo != null && target.ordinal() < combo.getOptions().size()) {
            combo.setOption(combo.getOptions().get(target.ordinal()));
        }
    }

    public void setScreenshotDisplayMode(ScreenshotDisplayMode mode) {
        ScreenshotDisplayMode target = mode == null ? ScreenshotDisplayMode.FILMSTRIP : mode;
        ComboSetting combo = getScreenshotDisplaySetting();
        if (combo != null && target.ordinal() < combo.getOptions().size()) {
            combo.setOption(combo.getOptions().get(target.ordinal()));
        }
    }

    public void setNotificationCorner(NotificationCorner corner) {
        NotificationCorner target = corner == null ? NotificationCorner.BOTTOM_RIGHT : corner;
        ComboSetting combo = getNotificationCornerSetting();
        if (combo != null && target.ordinal() < combo.getOptions().size()) {
            combo.setOption(combo.getOptions().get(target.ordinal()));
        }
    }

    public NumberSetting getVolumeSetting() {
        return SettingRegistry.getNumberSetting(this, "volumeSetting");
    }

    public ComboSetting getModThemeSetting() {
        return SettingRegistry.getComboSetting(this, "hudTheme");
    }

    public BooleanSetting getBlurSetting() {
        return SettingRegistry.getBooleanSetting(this, "blurSetting");
    }

    public KeybindSetting getModMenuKeybindSetting() {
        return SettingRegistry.getKeybindSetting(this, "modMenuKeybindSetting");
    }

    public BooleanSetting getMCHUDFont() {
        return SettingRegistry.getBooleanSetting(this, "mcFontSetting");
    }

    public BooleanSetting getBorderlessFullscreenSetting() {
        return SettingRegistry.getBooleanSetting(this, "borderlessFullscreenSetting");
    }

    public void applyBorderlessOnStartup() {
        borderlessInitialized = true;
        lastBorderlessState = borderlessFullscreenSetting;
        applyBorderlessSetting(borderlessFullscreenSetting, true);
    }

    private void applyBorderlessSetting(boolean state, boolean force) {
        if (!force && state == lastBorderlessState) {
            return;
        }

        lastBorderlessState = state;
        borderlessInitialized = true;

        if (!mc.isFullScreen()) {
            return;
        }

        if (state) {
            setBorderlessFullscreen(true);
        } else {
            setBorderlessFullscreen(false);
            mc.toggleFullscreen();
            mc.toggleFullscreen();
        }
    }

    private void setBorderlessFullscreen(boolean state) {
        try {
            System.setProperty("org.lwjgl.opengl.Window.undecorated", Boolean.toString(state));
            Display.setFullscreen(false);
            Display.setResizable(!state);

            if (state) {
                prevX = Display.getX();
                prevY = Display.getY();
                prevWidth = mc.displayWidth;
                prevHeight = mc.displayHeight;
                Display.setDisplayMode(new DisplayMode(Display.getDesktopDisplayMode().getWidth(), Display.getDesktopDisplayMode().getHeight()));
                Display.setLocation(0, 0);
                ((IMixinMinecraft) mc).resizeWindow(Display.getDesktopDisplayMode().getWidth(), Display.getDesktopDisplayMode().getHeight());
            } else {
                Display.setDisplayMode(new DisplayMode(prevWidth, prevHeight));
                Display.setLocation(prevX, prevY);
                ((IMixinMinecraft) mc).resizeWindow(prevWidth, prevHeight);

                if (mc.inGameHasFocus) {
                    mc.mouseHelper.ungrabMouseCursor();
                    fullscreenTime = System.currentTimeMillis();
                }
            }
        } catch (LWJGLException error) {
            ShindoLogger.error("Could not toggle borderless fullscreen", error);
        }
    }

    public String getCapeConfigName() {
        return capeNameSetting;
    }

    public void setCapeConfigName(String a) {
        capeNameSetting = a;
    }


    public enum HudTheme implements PropertyEnum {
        NORMAL(TranslateText.NORMAL),
        GLOW(TranslateText.GLOW),
        OUTLINE(TranslateText.OUTLINE),
        VANILLA(TranslateText.VANILLA),
        OUTLINE_GLOW(TranslateText.OUTLINE_GLOW),
        VANILLA_GLOW(TranslateText.VANILLA_GLOW),
        SHADOW(TranslateText.SHADOW),
        DARK(TranslateText.DARK),
        LIGHT(TranslateText.LIGHT),
        RECT(TranslateText.RECT),
        MODERN(TranslateText.MODERN),
        TEXT(TranslateText.TEXT),
        GRADIENT_SIMPLE(TranslateText.GRADIENT_SIMPLE);

        private final TranslateText translate;

        HudTheme(TranslateText translate) {
            this.translate = translate;
        }

        @Override
        public TranslateText getTranslate() {
            return translate;
        }

        @Override
        @Nullable
        public String getNameKey() {
            return PropertyEnum.super.getNameKey();
        }

        @Override
        @Nullable
        public String getDisplayName() {
            return PropertyEnum.super.getDisplayName();
        }
    }

    public enum SettingsLayout implements PropertyEnum {
        SINGLE_COLUMN("Single Column"),
        COMPACT_GRID("Compact Grid");

        private final String displayName;

        SettingsLayout(String displayName) {
            this.displayName = displayName;
        }

        @Override
        @Nullable
        public String getDisplayName() {
            return displayName;
        }

    }

    public enum ModuleLayout implements PropertyEnum {
        SINGLE_COLUMN("Single Column"),
        TWO_COLUMNS("Two Columns"),
        ICON_CARDS("Icon Cards");

        private final String displayName;

        ModuleLayout(String displayName) {
            this.displayName = displayName;
        }

        @Override
        @Nullable
        public String getDisplayName() {
            return displayName;
        }
    }

    public enum AddonLayout implements PropertyEnum {
        STANDARD("Standard"),
        ICON_CARDS("Icon Cards");

        private final String displayName;

        AddonLayout(String displayName) {
            this.displayName = displayName;
        }

        @Override
        @Nullable
        public String getDisplayName() {
            return displayName;
        }
    }

    public enum NotificationCorner implements PropertyEnum {
        TOP_LEFT(TranslateText.NOTIFICATION_POSITION_TOP_LEFT),
        TOP_RIGHT(TranslateText.NOTIFICATION_POSITION_TOP_RIGHT),
        BOTTOM_LEFT(TranslateText.NOTIFICATION_POSITION_BOTTOM_LEFT),
        BOTTOM_RIGHT(TranslateText.NOTIFICATION_POSITION_BOTTOM_RIGHT);

        private final TranslateText translate;

        NotificationCorner(TranslateText translate) {
            this.translate = translate;
        }

        @Override
        public TranslateText getTranslate() {
            return translate;
        }
    }
}

File: src/main/kotlin/me/miki/shindo/management/mods/impl/InternalSettingsMod.kt
package me.miki.shindo.management.mods.impl

import me.miki.shindo.Shindo
import me.miki.shindo.injection.interfaces.IMixinMinecraft
import me.miki.shindo.logger.ShindoLogger
import me.miki.shindo.management.event.EventTarget
import me.miki.shindo.management.event.impl.EventKey
import me.miki.shindo.management.event.impl.EventPreRenderTick
import me.miki.shindo.management.event.impl.EventToggleFullscreen
import me.miki.shindo.management.language.TranslateText
import me.miki.shindo.management.mods.Mod
import me.miki.shindo.management.mods.ModCategory
import me.miki.shindo.management.nanovg.font.LegacyIcon
import me.miki.shindo.management.settings.config.Property
import me.miki.shindo.management.settings.config.PropertyEnum
import me.miki.shindo.management.settings.config.PropertyType
import me.miki.shindo.management.settings.impl.BooleanSetting
import me.miki.shindo.management.settings.impl.ComboSetting
import me.miki.shindo.management.settings.impl.KeybindSetting
import me.miki.shindo.management.settings.impl.NumberSetting
import me.miki.shindo.management.settings.metadata.SettingRegistry.getBooleanSetting
import me.miki.shindo.management.settings.metadata.SettingRegistry.getComboSetting
import me.miki.shindo.management.settings.metadata.SettingRegistry.getKeybindSetting
import me.miki.shindo.management.settings.metadata.SettingRegistry.getNumberSetting
import me.miki.shindo.ui.animation.v2.GlobalAnimationSettings
import me.miki.shindo.ui.components.v2.layout.SettingsPanel.LayoutMode
import org.lwjgl.LWJGLException
import org.lwjgl.input.Keyboard
import org.lwjgl.opengl.Display
import org.lwjgl.opengl.DisplayMode
import java.util.*
import kotlin.math.max
import kotlin.math.min

class InternalSettingsMod : Mod(TranslateText.NONE, TranslateText.NONE, ModCategory.OTHER, LegacyIcon.MOD_INTERNAL_SETTINGS) {
    @Property(type = PropertyType.COMBO, translate = TranslateText.HUD_THEME)
    @JvmField
    val hudTheme: HudTheme = HudTheme.NORMAL

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.UI_BLUR)
    @JvmField
    var blurSetting = false

    @Property(
        type = PropertyType.NUMBER,
        translate = TranslateText.BLUR_STRENGTH,
        min = 10.0,
        max = 20.0,
        current = 1.0,
    )
    @JvmField
    var blurStrengthSetting = 1.0

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.ANIMATION)
    @JvmField
    var animationsSetting = true

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.MC_FONT)
    private val mcFontSetting = false

    @Property(type = PropertyType.NUMBER, translate = TranslateText.VOLUME, min = 0.0, max = 1.0, current = 0.8)
    @JvmField
    var volumeSetting = 0.8

    @Property(type = PropertyType.KEYBIND, translate = TranslateText.KEYBIND, keyCode = Keyboard.KEY_RSHIFT)
    private val modMenuKeybindSetting = Keyboard.KEY_RSHIFT

    @Property(type = PropertyType.TEXT, translate = TranslateText.CUSTOM_CAPE, text = "None")
    var capeConfigName: String? = "None"

    @Property(type = PropertyType.TEXT, name = "Custom Wing", text = "None")
    var wingConfigName: String? = "None"

    @Property(type = PropertyType.TEXT, name = "Custom Bandana", text = "None")
    var bandanaConfigName: String? = "None"

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.CLICK_EFFECT)
    @JvmField
    var clickEffectsSetting = false

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.UI_SOUNDS)
    @JvmField
    var soundsUISetting = false

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.BORDERLESS_FULSCREEN)
    @JvmField
    val borderlessFullscreenSetting = false

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.PERFORMANCE_TEXTURE_OPTIMIZATION)
    @JvmField
    var textureOptimizationSetting = true

    @Property(type = PropertyType.COMBO, name = "Settings Layout")
    val settingsLayout: SettingsLayout = SettingsLayout.SINGLE_COLUMN

    @Property(type = PropertyType.COMBO, name = "Visual Preset")
    @JvmField
    val visualPreset: VisualPreset = VisualPreset.MODERN

    @Property(type = PropertyType.COMBO, name = "Module Layout")
    @JvmField
    val moduleLayout = ModuleLayout.SINGLE_COLUMN

    @Property(type = PropertyType.COMBO, translate = TranslateText.NOTIFICATION_POSITION)
    var notificationCorner: NotificationCorner = NotificationCorner.BOTTOM_RIGHT
        get() = readEnumFromCombo(notificationCornerSetting, NotificationCorner.values(), field)
        set(corner) {
            writeEnumToCombo(this.notificationCornerSetting, corner)
            field = corner
        }

    private var prevX = 0
    private var prevY = 0
    private var prevWidth = 0
    private var prevHeight = 0
    private var fullscreenTime: Long = -1
    private var lastBorderlessState = false
    private var borderlessInitialized = false
    private var lastModMenuOpenTime = 0L

    init {
        instance = this
    }

    override fun setup() {
        this.setHide(true)
        this.setToggled(true)
    }

    @EventTarget
    fun onKey(event: EventKey) {
        val keybind = getModMenuKeybindSetting()?.getKeyCode() ?: modMenuKeybindSetting
        if (event.keyCode == keybind) {
            mc.displayGuiScreen(Shindo.getInstance().getShindoAPI().navigationHub)
        }

        if (event.keyCode == Keyboard.KEY_DOWN) {
            val combo = this.modThemeSetting
            if (combo != null) {
                val max = combo.getOptions().size
                var modeIndex = combo.getOptions().indexOf(combo.getOption())
                modeIndex = if (modeIndex > 0) modeIndex - 1 else max - 1
                combo.setOption(combo.getOptions()[modeIndex])
            }
        }
    }

    @EventTarget
    fun onRenderTick(event: EventPreRenderTick?) {
        GlobalAnimationSettings.enabled = animationsSetting
        if (fullscreenTime != -1L && System.currentTimeMillis() - fullscreenTime >= 100) {
            fullscreenTime = -1

            if (mc.inGameHasFocus) {
                mc.mouseHelper.grabMouseCursor()
            }
        }

        if (borderlessInitialized && borderlessFullscreenSetting != lastBorderlessState) {
            applyBorderlessSetting(borderlessFullscreenSetting, false)
        }
    }

    @EventTarget
    fun onFullscreenToggle(event: EventToggleFullscreen) {
        if (!borderlessFullscreenSetting) {
            return
        }
        event.isApplyState = false
        setBorderlessFullscreen(event.state)
    }

    fun getModuleLayout(): ModuleLayout = readEnumFromCombo(moduleLayoutSetting, ModuleLayout.values(), moduleLayout)

    private val settingsLayoutSetting: ComboSetting?
        get() = getComboSetting(this, "settingsLayout")

    private val visualPresetSetting: ComboSetting?
        get() = getComboSetting(this, "visualPreset")

    private val moduleLayoutSetting: ComboSetting?
        get() = getComboSetting(this, "moduleLayout")

    private val notificationCornerSetting: ComboSetting?
        get() = getComboSetting(this, "notificationCorner")

    var settingsLayoutMode: LayoutMode?
        get() =
            when (readEnumFromCombo(settingsLayoutSetting, SettingsLayout.values(), settingsLayout)) {
                SettingsLayout.DOUBLE_COLUMN -> LayoutMode.DOUBLE_COLUMN
                SettingsLayout.ADAPTIVE_GRID -> LayoutMode.STAGGERED_COLUMNS
                else -> LayoutMode.SINGLE_COLUMN
            }
        set(mode) {
            val target =
                when (mode) {
                    LayoutMode.DOUBLE_COLUMN -> SettingsLayout.DOUBLE_COLUMN
                    LayoutMode.STAGGERED_COLUMNS -> SettingsLayout.ADAPTIVE_GRID
                    else -> SettingsLayout.SINGLE_COLUMN
                }
            writeEnumToCombo(this.settingsLayoutSetting, target)
        }

    var moduleGridColumns: Int
        get() {
            if (Objects.requireNonNull<ModuleLayout?>(getModuleLayout()) == ModuleLayout.TWO_COLUMNS) {
                return 2
            }
            return 1
        }
        set(columns) {
            val normalized = max(1, min(columns, 2))
            val target = if (normalized == 2) ModuleLayout.TWO_COLUMNS else ModuleLayout.SINGLE_COLUMN
            writeEnumToCombo(this.moduleLayoutSetting, target)
        }

    fun getVisualPreset(): VisualPreset = readEnumFromCombo(visualPresetSetting, VisualPreset.values(), visualPreset)

    fun setVisualPreset(preset: VisualPreset?) {
        val target = preset ?: VisualPreset.MODERN
        writeEnumToCombo(this.visualPresetSetting, target)
    }

    fun setModuleLayout(layout: ModuleLayout?) {
        val target = layout ?: ModuleLayout.SINGLE_COLUMN
        writeEnumToCombo(this.moduleLayoutSetting, target)
    }

    val modThemeSetting: ComboSetting?
        get() = getComboSetting(this, "hudTheme")

    fun getModMenuKeybindSetting(): KeybindSetting? = getKeybindSetting(this, "modMenuKeybindSetting")

    val mCHUDFont: BooleanSetting?
        get() = getBooleanSetting(this, "mcFontSetting")

    fun getBorderlessFullscreenSetting(): BooleanSetting? = getBooleanSetting(this, "borderlessFullscreenSetting")

    fun getBlurSetting(): BooleanSetting? = getBooleanSetting(this, "blurSetting")

    fun getBlurStrengthSetting(): NumberSetting? = getNumberSetting(this, "blurStrengthSetting")

    fun getAnimationsSetting(): BooleanSetting? = getBooleanSetting(this, "animationsSetting")

    fun getVolumeSetting(): NumberSetting? = getNumberSetting(this, "volumeSetting")

    fun getClickEffectsSetting(): BooleanSetting? = getBooleanSetting(this, "clickEffectsSetting")

    fun getSoundsUISetting(): BooleanSetting? = getBooleanSetting(this, "soundsUISetting")

    fun getTextureOptimizationSetting(): BooleanSetting? = getBooleanSetting(this, "textureOptimizationSetting")

    private fun <T : Enum<T>> readEnumFromCombo(
        combo: ComboSetting?,
        values: Array<T>,
        fallback: T,
    ): T {
        if (combo == null) {
            return fallback
        }
        val selected = combo.getOption() ?: return fallback
        val index = combo.getOptions().indexOf(selected)
        return if (index >= 0 && index < values.size) values[index] else fallback
    }

    private fun <T : Enum<T>> writeEnumToCombo(
        combo: ComboSetting?,
        value: T,
    ) {
        if (combo == null) {
            return
        }
        if (value.ordinal >= 0 && value.ordinal < combo.getOptions().size) {
            combo.setOption(combo.getOptions()[value.ordinal])
        }
    }

    fun applyBorderlessOnStartup() {
        borderlessInitialized = true
        lastBorderlessState = borderlessFullscreenSetting
        applyBorderlessSetting(borderlessFullscreenSetting, true)
    }

    private fun applyBorderlessSetting(
        state: Boolean,
        force: Boolean,
    ) {
        if (!force && state == lastBorderlessState) {
            return
        }

        lastBorderlessState = state
        borderlessInitialized = true

        if (!mc.isFullScreen) {
            return
        }

        if (state) {
            setBorderlessFullscreen(true)
        } else {
            setBorderlessFullscreen(false)
            mc.toggleFullscreen()
            mc.toggleFullscreen()
        }
    }

    private fun setBorderlessFullscreen(state: Boolean) {
        try {
            System.setProperty("org.lwjgl.opengl.Window.undecorated", state.toString())
            Display.setFullscreen(false)
            Display.setResizable(!state)

            if (state) {
                prevX = Display.getX()
                prevY = Display.getY()
                prevWidth = mc.displayWidth
                prevHeight = mc.displayHeight
                Display.setDisplayMode(
                    DisplayMode(
                        Display.getDesktopDisplayMode().width,
                        Display.getDesktopDisplayMode().height,
                    ),
                )
                Display.setLocation(0, 0)
                (mc as IMixinMinecraft).resizeWindow(
                    Display.getDesktopDisplayMode().width,
                    Display.getDesktopDisplayMode().height,
                )
            } else {
                Display.setDisplayMode(DisplayMode(prevWidth, prevHeight))
                Display.setLocation(prevX, prevY)
                (mc as IMixinMinecraft).resizeWindow(prevWidth, prevHeight)

                if (mc.inGameHasFocus) {
                    mc.mouseHelper.ungrabMouseCursor()
                    fullscreenTime = System.currentTimeMillis()
                }
            }
        } catch (error: LWJGLException) {
            ShindoLogger.error("Could not toggle borderless fullscreen", error)
        }
    }

    enum class HudTheme(
        private val translate: TranslateText,
    ) : PropertyEnum {
        NORMAL(TranslateText.NORMAL),
        GLOW(TranslateText.GLOW),
        OUTLINE(TranslateText.OUTLINE),
        VANILLA(TranslateText.VANILLA),
        OUTLINE_GLOW(TranslateText.OUTLINE_GLOW),
        VANILLA_GLOW(TranslateText.VANILLA_GLOW),
        SHADOW(TranslateText.SHADOW),
        DARK(TranslateText.DARK),
        LIGHT(TranslateText.LIGHT),
        RECT(TranslateText.RECT),
        MODERN(TranslateText.MODERN),
        TEXT(TranslateText.TEXT),
        GRADIENT_SIMPLE(TranslateText.GRADIENT_SIMPLE),
        ;

        override fun getTranslate(): TranslateText = translate
    }

    enum class SettingsLayout(
        displayName: String,
    ) : PropertyEnum {
        SINGLE_COLUMN("Single Column"),
        DOUBLE_COLUMN("Double Column"),
        ADAPTIVE_GRID("Adaptive Grid"),
        ;

        private val displayName: String = displayName

        override fun getDisplayName(): String = displayName
    }

    enum class ModuleLayout(
        displayName: String,
    ) : PropertyEnum {
        SINGLE_COLUMN("Single Column"),
        TWO_COLUMNS("Two Columns"),
        ;

        private val displayName: String = displayName

        override fun getDisplayName(): String = displayName
    }

    enum class VisualPreset(
        displayName: String,
    ) : PropertyEnum {
        CLASSIC("Classic"),
        MODERN("Modern"),
        LIGHT("Light"),
        DARK("Dark"),
        ;

        private val displayName: String = displayName

        override fun getDisplayName(): String = displayName
    }

    enum class NotificationCorner(
        private val translate: TranslateText,
    ) : PropertyEnum {
        TOP_LEFT(TranslateText.NOTIFICATION_POSITION_TOP_LEFT),
        TOP_RIGHT(TranslateText.NOTIFICATION_POSITION_TOP_RIGHT),
        BOTTOM_LEFT(TranslateText.NOTIFICATION_POSITION_BOTTOM_LEFT),
        BOTTOM_RIGHT(TranslateText.NOTIFICATION_POSITION_BOTTOM_RIGHT),
        ;

        override fun getTranslate(): TranslateText = translate
    }

    companion object {
        lateinit var instance: InternalSettingsMod
    }
}
